src/fsm.c

```diff
@@ -61,6 +61,123 @@ static int add_pending_pages(struct fsm *f,
 	return 0;
 }
 
+static int databaseReadLock(struct db *db)
+{
+    if (!db->read_lock) {
+	    db->read_lock = 1;
+	    return 0;
+    } else {
+	    return -1;
+    }
+}
+
+static int databaseReadUnlock(struct db *db)
+{
+    if (db->read_lock) {
+	    db->read_lock = 0;
+	    return 0;
+    } else {
+	    return -1;
+    }
+}
+
+static void maybeCheckpoint(struct db *db)
+{
+	tracef("maybe checkpoint");
+	struct sqlite3_file *main_f;
+	struct sqlite3_file *wal;
+	volatile void *region;
+	sqlite3_int64 size;
+	unsigned page_size;
+	unsigned pages;
+	int wal_size;
+	int ckpt;
+	int i;
+	int rv;
+
+	/* Don't run when a snapshot is busy. Running a checkpoint while a snapshot
+	 * is busy will result in illegal memory accesses by the routines that try
+	 * to access database page pointers contained in the snapshot. */
+	rv = databaseReadLock(db);
+	if (rv != 0) {
+		tracef("busy snapshot %d", rv);
+		return;
+	}
+
+	assert(db->follower == NULL);
+	rv = db__open_follower(db);
+	if (rv != 0) {
+		tracef("open follower failed %d", rv);
+		goto unlock;
+	}
+
+	page_size = db->config->page_size;
+	/* Get the database wal file associated with this connection */
+	rv = sqlite3_file_control(db->follower, "main", SQLITE_FCNTL_JOURNAL_POINTER,
+				  &wal);
+	assert(rv == SQLITE_OK); /* Should never fail */
+
+	rv = wal->pMethods->xFileSize(wal, &size);
+	assert(rv == SQLITE_OK); /* Should never fail */
+
+	/* Calculate the number of frames. */
+	pages = ((unsigned)size - 32) / (24 + page_size);
+
+	/* Check if the size of the WAL is beyond the threshold. */
+	if (pages < db->config->checkpoint_threshold) {
+		tracef("wal size (%u) < threshold (%u)", pages, db->config->checkpoint_threshold);
+		goto cleanup;
+	}
+
+	/* Get the database file associated with this db->follower connection */
+	rv = sqlite3_file_control(db->follower, "main", SQLITE_FCNTL_FILE_POINTER,
+				  &main_f);
+	assert(rv == SQLITE_OK); /* Should never fail */
+
+	/* Get the first SHM region, which contains the WAL header. */
+	rv = main_f->pMethods->xShmMap(main_f, 0, 0, 0, &region);
+	assert(rv == SQLITE_OK); /* Should never fail */
+
+	rv = main_f->pMethods->xShmUnmap(main_f, 0);
+	assert(rv == SQLITE_OK); /* Should never fail */
+
+	/* Try to acquire all locks. */
+	for (i = 0; i < SQLITE_SHM_NLOCK; i++) {
+		int flags = SQLITE_SHM_LOCK | SQLITE_SHM_EXCLUSIVE;
+
+		rv = main_f->pMethods->xShmLock(main_f, i, 1, flags);
+		if (rv == SQLITE_BUSY) {
+			tracef("busy reader or writer - retry next time");
+			goto cleanup;
+		}
+
+		/* Not locked. Let's release the lock we just
+		 * acquired. */
+		flags = SQLITE_SHM_UNLOCK | SQLITE_SHM_EXCLUSIVE;
+		main_f->pMethods->xShmLock(main_f, i, 1, flags);
+	}
+
+	rv = sqlite3_wal_checkpoint_v2(
+	     db->follower, "main", SQLITE_CHECKPOINT_TRUNCATE, &wal_size, &ckpt);
+	if (rv != 0) {
```

Comment:
I would have expected this to be an `assert(rv == 0)`, but I noticed the old code removed just below does the same. It might be interesting to dig in git history to see if it has always been like this or was added for a specific reason. If it was always like this, perhaps it'd be worth checking the SQLite code if there are actually valid failure modes that we have to guard from.

This is optional, since basically it doesn't change current behavior.

Reply:
Thanks, I'll add a `TODO` for now.